### DISCUSSION TITLE
Is it possible to allow implicit conversion from Python lists to ndarray?

### DISCUSSION THREAD

**Discussion by Ahajha** [wjakob/nanobind#327] [Q&A]
In porting some existing pybind11 code, it seems pybind11::array_t would accept a list (emphasis on seems, I did not write said code). I find myself having to provide a second version of everything taking an ndarray where I would also want to allow a list to be passed in. It's fine if this is the case, but is there a way of allowing such an implicit conversion?

**Comment by wjakob** [comment]
That's a bit more tricky in the case of nanobind because it has no dependence on NumPy for ``array_t``. So it wouldn't be reasonable to first call ``np.array()`` and then convert from that. You would need a custom path that internally converts the list into a DLPack tensor and then passes that to the existing dlpack tensor interface.

I'm open to adding such a feature, but it's not high on my own priority list. Would you be interested in adding something like that?

**Reply by Ahajha** [reply]
I could maybe look into implementing it. Personally I'm only interested in simple 1D cases (such as [1, 2, 3]), but I can see use for the ND case as well. Perhaps a `def` flag that can be passed in? `nb::allow_list_to_ndarray`? Or maybe just always allow the conversion (adding an implicit conversion)?

**Reply by wjakob** [reply]
I'd always allow the implicit conversion.

**Reply by Ahajha** [reply]
I'm poking around in the implementation in some spare time, would you mind if I made an issue out of this discussion? I know you like to keep the issue list clean, so I figured I'd check with you first.

**Reply by wjakob** [reply]
If you have tentative code, then how about you put it into a PR and we continue the discussion there? I prefer to keep 'Issues' limited to things that are actual bugs (having a nonzero number there always stresses me out 🥲).

**Reply by Ahajha** [reply]
Nothing yet, I basically have figured out:
1. Where the code should go
2. A sketch of the algorithm to convert "NDlists" to ndarrays

Hopefully I'll have something to put into a draft PR maybe in a week or so?

**Reply by Ahajha** [reply]
Been working on other things (CPython package for Conan 2.0, which also unblocks a nanobind Conan package), but I think I can start giving this some attention again.

**Reply by jmalkin** [reply]
@Ahajha I don't think I know the internals of nanobind well enough to help, but I just ran into a situation where this addition would be quite useful.

**Reply by Ahajha** [reply]
That's good to hear, gives me a little bit more motivation to work on it. I'm probably a month or two until I really need it, but I'm working on it here and there. Hopefully soon I'll have a (very) rough draft that I can put into a PR.

**Reply by wojdyr** [reply]
I could use that as well. I wonder, is anyone looking into it currently?

**Reply by Ahajha** [reply]
I probably should have said something earlier, I am no longer working on the above project, and I unfortunately haven't had the time to work on this feature. So unless someone else is, I believe the answer is that nobody is working on this.

**Reply by eudoxos** [reply]
I might get to this, asking for general guidance. `ndarray_import ` does currently (https://github.com/wjakob/nanobind/blob/62fc996018d9ea4d51af9c86cf008c2562b4eeab/src/nb_ndarray.cpp#L381-421) 

1. check for dltensor capsule;
2. if that fails, check for known tensor packages (tensorflow, toch, jaxlib) and call their `to_dlpack` method;
3. if that fails, attempt `dlpack_from_buffer_protocol`

There would be an extra step added after that:

4. if that fails, attempt `dlpack_from_sequence_protocol` (to be implemented).

This new function would receive expected `ndarray_config` so that it can fail early when the input is not compatible (mismatched dimension, shape, unequal lengths of rows in a matrix, incompatible element type (`int` would be valid for integral dtypes, `int || float` for floating dtypes, complex for complex dtypes; `int<0` would be refused for unsigned dtypes). 

Like that?

**Reply by wjakob** [reply]
One question is how comprehensive `dlpack_from_sequence_protocol` should be. If the function expects a CUDA tensor, it is not realistic for nanobind to link to the CUDA library to move memory to the GPU to satisfy the requirements. The best we can do is to make a CPU tensor. And even that is quite complicated (e.g., figuring out the required type in multi-dimensional arrays, where maybe most of the entries are `int` and one of them is `float`....)

A less efficient but technically far simpler solution could be used in the case where the `nb::ndarray<>` type declaration is annotated with a given framework template parameter (e.g. `nb::numpy`). In that case the simplest solution would be to recursively call NumPy (or whatever framework was listed) to perform an implicit conversion from a sequence to the associated tensor type and then cast that using the available methods.

**Reply by eudoxos** [reply]
Thank you for the input. What I propose (simple, not to get overwhelmed):

1. restrict these conversions to numpy only (at first, at least) —  I would assume GPU computations will rarely have small hand-written arrays like [1,2,3] as function parameters in the code (numpy is my use-case, so I might be biased).
2. enable the attempted conversion via `#ifdef NPY_*` (something, TBD), meaning nanobind will not have to deal with actually including numpy headers, as they have been included already.
3. Only attempt the conversion if the array is annotated as `nb::numpy` (or also if unannotated?).
4. Attempt to call [PyArray_FromAny](https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_FromAny) which will do all the heavylifting like nested sequences parsing, type conversions etc for us. Return either dlpack based on the array constructed (to be consistent with what is returned otherwise), or just numpy array as object already.
5. If the conversion fails, continue as usual.

**Reply by wjakob** [reply]
Including the NumPy headers is not an option (even if optional), so everything will need to go through the Python interface (which should be possible). To consider such a feature for inclusion in nanobind, it should also work on the other array frameworks.
